### PR TITLE
udev-watch: retry to save watch handle on error

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/udev/udev-watch.c [hunk 1]
```diff
@@ -12,8 +12,13 @@
 #include "dirent-util.h"
 #include "fs-util.h"
 #include "parse-util.h"
+#include "random-util.h"
 #include "udev-watch.h"
 
+#define SAVE_WATCH_HANDLE_MAX_RETRIES  128
+#define MAX_RANDOM_DELAY (250 * USEC_PER_MSEC)
+#define MIN_RANDOM_DELAY ( 50 * USEC_PER_MSEC)
```
Comment (bluca) [inline src/udev/udev-watch.c]: In some cases 50ms is a lot and measurable when profiling, can this be a bit smaller? 10?

Comment (yuwata) [reply src/udev/udev-watch.c]: OK. Now, the range is 10-100 msec.

File: src/udev/udev-watch.c [hunk 2]
```diff
@@ -93,11 +98,43 @@ int udev_watch_begin(int inotify_fd, sd_device *dev) {
...
         }
 
-        r = device_set_watch_handle(dev, wd);
-        if (r < 0)
-                return log_device_warning_errno(dev, r, "Failed to save watch handle in /run/udev/watch: %m");
+        for (unsigned i = 0; i < SAVE_WATCH_HANDLE_MAX_RETRIES; i++) {
+                if (i > 0) {
+                        usec_t delay = MIN_RANDOM_DELAY + random_u64_range(MAX_RANDOM_DELAY - MIN_RANDOM_DELAY);
+
+                        /* When the same handle is reused for different device node, we may failed to
```
Comment (bluca) [inline src/udev/udev-watch.c]: "we may fail"

File: src/udev/udev-watch.c [hunk 3]
```diff
@@ -93,11 +98,43 @@ int udev_watch_begin(int inotify_fd, sd_device *dev) {
...
+                         * save the watch handle with -EEXIST. Let's consider the case two workers A
```
Comment (bluca) [inline src/udev/udev-watch.c]: "of two workers"

File: src/udev/udev-watch.c [hunk 4]
```diff
@@ -93,11 +98,43 @@ int udev_watch_begin(int inotify_fd, sd_device *dev) {
...
+                         * 2. B calls inotify_add_watch()
+                         * 3. B calls device_set_watch_handle()
+                         * 4. A calls device_set_watch_handle(-1)
+                         *
+                         * At step 3, the old symlinks to save the watch handle still exist. So,
+                         * device_set_watch_handle() fails with -EEXIST. */
+
+                        log_device_debug_errno(dev, r,
+                                               "Failed to save watch handle '%i' for %s in "
+                                               "/run/udev/watch, retrying after %s: %m",
```
Comment (bluca) [inline src/udev/udev-watch.c]: "retrying in"

**Review by bluca — APPROVED** [review]
lgtm, just some nits

**Comment by yuwata** [comment]
@bluca Thank you for your review (and sorry for my bad English...). Updated.

**Comment by poettering** [comment]
Can you explain when this can happen? i.e. how could it be that wd or device id conflict?

Or is this simply a race, because wds and device ids are reused?

**Comment by poettering** [comment]
if this is about fixing the race, then this lgtm

**Comment by yuwata** [comment]
This fixes a race when a watch handle is reused. As I commented in the code, when two workers A and B do the following,

1. A calls inotify_rm_watch()
2. B calls inotify_add_watch()
3. B calls device_set_watch_handle()
4. A calls device_set_watch_handle(-1)

then step 3 fails if the watch handle is reused, that is, the same watch handle removed in step 1 is obtained in step 2.

**Pull request merged**